File: go/dotprompt/test/spec_test.go
// Copyright 2025 Google LLC
// SPDX-License-Identifier: Apache-2.0

package dotprompt

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	"github.com/go-viper/mapstructure/v2"
	. "github.com/google/dotprompt/go/dotprompt"
	"github.com/invopop/jsonschema"
)

const SpecDir = "../../../spec"

func TestSpecFiles(t *testing.T) {
	processSpecFiles(t)
}

// compareMaps performs a deep comparison of two maps of type map[string]any.
func compareMaps(map1, map2 map[string]any) bool {
	if len(map1) != len(map2) {
		return false
	}
	for k, v1 := range map1 {
		v2, ok := map2[k]
		if !ok {
			return false
		}
		if !deepEqual(v1, v2) {
			return false
		}
	}
	return true
}

// deepEqual performs a deep comparison of two values.
func deepEqual(v1, v2 any) bool {
	if reflect.TypeOf(v1) != reflect.TypeOf(v2) {
		return false
	}
	switch v1 := v1.(type) {
	case map[string]any:
		v2, ok := v2.(map[string]any)
		if !ok {
			return false
		}
		return compareMaps(v1, v2)
	case []any:
		v2, ok := v2.([]any)
		if !ok {
			return false
		}
		if len(v1) != len(v2) {
			return false
		}
		for i := range v1 {
			if !deepEqual(v1[i], v2[i]) {
				return false
			}
		}
		return true
	default:
		return reflect.DeepEqual(v1, v2)
	}
}

// createTestCases creates and runs test cases for a given SpecSuite and SpecTest.
func createTestCases(t *testing.T, s SpecSuite, tc SpecTest, dotpromptFactory func(suite SpecSuite) (*Dotprompt, *DotpromptOptions)) {
	t.Run(tc.Desc, func(t *testing.T) {
		env, dotpromptOptions := dotpromptFactory(s)

		// Render the template.
		options := &PromptMetadata{}
		if err := mapstructure.Decode(tc.Options, options); err != nil {
			t.Fatalf("Failed to decode options: %v", err)
		}
		dataArg := mergeData(s.Data, tc.Data)
		result, err := env.Render(s.Template, &dataArg, options, dotpromptOptions)
		if err != nil {
			t.Fatalf("Render failed: %v", err)
		}
		// Prune the result and compare to the expected output.
		prunedResult := pruneResult(t, result.PromptMetadata)
		if len(result.Messages) > 0 {
			prunedResult["messages"] = pruneMessages(result.Messages)
		}
		expected := pruneExpected(tc.Expect)

		// Compare the pruned result to the expected output.
		if !compareResults(prunedResult, expected) {
			t.Errorf("Render should produce the expected result. Got: %v, Expected: %v", prunedResult, expected)
		}

		// Only compare raw if the spec demands it.
		if tc.Expect.Raw != nil {
			if !compareMaps(result.Raw, tc.Expect.Raw) {
				t.Errorf("Raw output mismatch. Got: %v, Expected: %v", result.Raw, tc.Expect.Raw)
			}
		}
	})
}

// createTestSuite creates and runs a test suite for a given suite name and SpecSuite.
func createTestSuite(t *testing.T, suiteName string, suites []SpecSuite, dotpromptFactory func(suite SpecSuite) (*Dotprompt, *DotpromptOptions)) {
	t.Run(suiteName, func(t *testing.T) {
		for _, s := range suites {
			t.Run(s.Name, func(t *testing.T) {
				for _, tc := range s.Tests {
					createTestCases(t, s, tc, dotpromptFactory)
				}
			})
		}
	})
}

// processSpecFile processes a single spec file and creates a test suite for it.
func processSpecFile(t *testing.T, file string, dotpromptFactory func(suite SpecSuite) (*Dotprompt, *DotpromptOptions)) {
	suiteName := filepath.Base(file)
	content, err := os.ReadFile(file)
	if err != nil {
		t.Fatalf("Failed to read file: %v", err)
	}
	fmt.Println(suiteName)
	suites := convertToSpecSuite(t, content)
	createTestSuite(t, suiteName, suites, dotpromptFactory)
}

// processSpecFiles processes all spec files in the SpecDir directory.
func processSpecFiles(t *testing.T) {
	err := filepath.Walk(SpecDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if !info.IsDir() && filepath.Ext(info.Name()) == ".yaml" {
			processSpecFile(t, path, func(s SpecSuite) (*Dotprompt, *DotpromptOptions) {
				options := &DotpromptOptions{
					Schemas:  s.Schemas,
					Tools:    s.Tools,
					Partials: s.Partials,
					PartialResolver: func(name string) (string, error) {
						if partial, ok := s.ResolverPartials[name]; ok {
							return partial, nil
						}
						return "", nil
					},
				}
				return NewDotprompt(options), options
			})
		}
		return nil
	})
	if err != nil {
		t.Fatalf("Failed to read spec directory: %v", err)
	}
}

// mergeData merges two DataArgument objects.
func mergeData(data1, data2 DataArgument) DataArgument {
	merged := data1
	if data2.Input != nil {
		if merged.Input == nil {
			merged.Input = make(map[string]any)
		}
		for k, v := range data2.Input {
			merged.Input[k] = v
		}
	}
	if data2.Docs != nil {
		merged.Docs = append(merged.Docs, data2.Docs...)
	}
	if data2.Messages != nil {
		merged.Messages = append(merged.Messages, data2.Messages...)
	}
	if data2.Context != nil {
		if merged.Context == nil {
			merged.Context = make(map[string]any)
		}
		for k, v := range data2.Context {
			merged.Context[k] = v
		}
	}
	return merged
}

// pruneResult prunes the result of a PromptMetadata object for comparison.
func pruneResult(t *testing.T, result PromptMetadata) map[string]any {
	pruned := make(map[string]any)
	if len(result.Config) > 0 {
		pruned["config"] = result.Config
	}
	if len(result.Ext) > 0 {
		pruned["ext"] = result.Ext
	}
	if result.Input.Default != nil || result.Input.Schema != nil {
		inputMap := make(map[string]any)
		if result.Input.Schema != nil {
			if inputSchema, ok := result.Output.Schema.(*jsonschema.Schema); ok {
				rawInput, _ := result.Raw["output"].(map[string]any)
				rawSchema, _ := rawInput["schema"].(map[string]any)
				inputMap["schema"] = pruneSchema(inputSchema, rawSchema)
			}
		}
		if result.Input.Default != nil {
			inputMap["default"] = result.Input.Default
		}
		pruned["input"] = inputMap
	}
	if result.Output.Format != "" || result.Output.Schema != nil {
		outputMap := make(map[string]any)
		if result.Output.Schema != nil {
			if outputSchema, ok := result.Output.Schema.(*jsonschema.Schema); ok {
				rawOutput, _ := result.Raw["output"].(map[string]any)
				rawSchema, _ := rawOutput["schema"].(map[string]any)
				outputMap["schema"] = pruneSchema(outputSchema, rawSchema)
			}
		}
		if result.Output.Format != "" {
			outputMap["format"] = result.Output.Format
		}
		pruned["output"] = outputMap
	}
	if len(result.HasMetadata.Metadata) > 0 {
		pruned["metadata"] = result.HasMetadata.Metadata
	}
	return pruned
}

// pruneExpected prunes the expected output for comparison.
func pruneExpected(expect Expect) map[string]any {
	pruned := make(map[string]any)
	if len(expect.Config) > 0 {
		pruned["config"] = expect.Config
	}
	if len(expect.Ext) > 0 {
		pruned["ext"] = expect.Ext
	}
	if len(expect.Input) > 0 {
		pruned["input"] = expect.Input
	}
	if len(expect.Output) > 0 {
		pruned["output"] = expect.Output
	}
	if len(expect.Messages) > 0 {
		pruned["messages"] = expect.Messages
	}
	if len(expect.Metadata) > 0 {
		pruned["metadata"] = expect.Metadata
	}
	return pruned
}

// pruneMessages prunes the messages for comparison.
func pruneMessages(messages []Message) []map[string]any {
	pruned := make([]map[string]any, 0)
	for _, message := range messages {
		prunedMessage := make(map[string]any)
		if len(message.Content) > 0 {
			prunedMessage["content"] = pruneContent(message.Content)
		}
		if len(message.HasMetadata.Metadata) > 0 {
			prunedMessage["metadata"] = message.HasMetadata.Metadata
		}
		if len(message.Role) > 0 {
			prunedMessage["role"] = message.Role
		}
		pruned = append(pruned, prunedMessage)
	}
	return pruned
}

// pruneContent prunes the content of a message for comparison.
func pruneContent(content []Part) []map[string]any {
	pruned := make([]map[string]any, 0)
	for _, part := range content {
		prunedPart := make(map[string]any)
		switch p := part.(type) {
		case *TextPart:
			if p.Text != "" {
				prunedPart["text"] = p.Text
			}
		case *DataPart:
			if len(p.Data) > 0 {
				prunedPart["data"] = p.Data
			}
		case *MediaPart:
			if p.Media.URL != "" || p.Media.ContentType != "" {
				prunedPart["media"] = map[string]any{
					"url":         p.Media.URL,
					"contentType": p.Media.ContentType,
				}
			}
		case *ToolRequestPart:
			if len(p.ToolRequest) > 0 {
				prunedPart["toolRequest"] = p.ToolRequest
			}
		case *ToolResponsePart:
			if len(p.ToolResponse) > 0 {
				prunedPart["toolResponse"] = p.ToolResponse
			}
		}
		if len(part.GetMetadata()) > 0 {
			prunedPart["metadata"] = part.GetMetadata()
		}
		pruned = append(pruned, prunedPart)
	}
	return pruned
}

func pruneSchema(schema *jsonschema.Schema, rawSchema map[string]any) map[string]any {
	schemaMap := make(map[string]any)
	if len(schema.AnyOf) != 0 {
		schemaMap["type"] = []string{}
		typeList := []string{}
		for _, anySchema := range schema.AnyOf {
			if anySchema.Type == "null" {
				typeList = append(typeList, "null")
			} else {
				typeList = append(typeList, anySchema.Type)
			}
		}
		schemaMap["type"] = typeList
	} else if schema.Type != "" {
		schemaMap["type"] = schema.Type
	}
	if schema.Description != "" {
		schemaMap["description"] = schema.Description
	}
	if schema.Items != nil {
		schemaMap["items"] = pruneSchema(schema.Items, rawSchema)
	}

	if schema.Type == "object" {
		if schema.AdditionalProperties == nil {
			if rawSchema != nil {
				schemaBytes, _ := json.Marshal(rawSchema)
				schemaJSON := &jsonschema.Schema{}
				_ = json.Unmarshal(schemaBytes, schemaJSON)
				// Validate that all fields in schemaMap are present in schemaJSON
				if err := ValidateSchemaFields(rawSchema, schemaJSON); err != nil {
					schemaMap["additionalProperties"] = false
				}
			}
		} else {
			schemaMap["additionalProperties"] = pruneSchema(schema.AdditionalProperties, rawSchema)
		}
		if len(schema.Required) != 0 {
			schemaMap["required"] = schema.Required
		}
		propMap := make(map[string]any)
		for property := schema.Properties.Oldest(); property != nil; property = property.Next() {
			propName := property.Key
			prop := property.Value
			propMap[propName] = pruneSchema(prop, rawSchema)
		}
		schemaMap["properties"] = propMap
	}

	if len(schema.Enum) != 0 {
		schemaMap["enum"] = schema.Enum
	}

	return schemaMap

}

// compareResults compares the result and expected output.
func compareResults(result, expected map[string]any) bool {
	resultJSON, _ := json.Marshal(result)
	expectedJSON, _ := json.Marshal(expected)
	return string(resultJSON) == string(expectedJSON)
}

File: python/dotpromptz/tests/dotpromptz/spec_test.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Runs specification tests defined in YAML files located in the `spec` directory.

These YAML files define test suites for Dotprompt templates.
The general structure of each YAML file is a list of test suites.

- Each YAML file is a set of test suites.
- Each test suite is a set of tests.
- Each test is a set of expectations.

Each test suite object in the list typically contains the following keys:

| Key                | Type   | Description                                                                          |
|--------------------|--------|--------------------------------------------------------------------------------------|
| `name`             | string | An identifier for this group of tests.                                               |
| `template`         | string | The Dotprompt template content (often multi-line, and may include YAML frontmatter). |
| `tests`            | list   | A list of individual test scenarios. Each scenario object has keys: `desc` (string), |
|                    |        | `data` (object, optional), `expect` (object).                                        |
| `partials`         | object | (Optional) A mapping where keys are partial names and values are the string content  |
|                    |        | of those partials. (e.g., see `partials.yaml`)                                       |
| `resolverPartials` | object | (Optional) Similar to `partials`, but for partials                                   |
|                    |        | that are expected to be provided by a resolver function. (e.g., see `partials.yaml`) |

Each test scenario object (i.e., an item in the `tests` list) has the following keys:

| Key      | Type   | Description                                                         |
|----------|--------|---------------------------------------------------------------------|
| `desc`   | string | A description of the specific test case.                            |
| `data`   | object | (Optional) Input data for the template. This can include `context`, |
|          |        | `input`, or other variables relevant to the template.               |
| `expect` | object | Defines the expected outcome. Common keys include                   |
|          |        | (list), and others corresponding to parsed frontmatter (`config`,   |
|          |        | `model`, `output.schema`, `input.schema`, `ext`, `raw`).            |

The YAML files located in the `spec/helpers` subdirectory also follow this
general structure, but are specifically focused on testing individual helper
functions available within the Dotprompt templating environment.

The structure of the YAML files is as follows:

```
spec/
├── *.yaml (e.g., metadata.yaml, partials.yaml, picoschema.yaml, etc.)
│   └── List of Test Suites
│       └── Test Suite 1
│           ├── name: "suite_name_1"
│           ├── template: "..."
│           ├── partials: { ... } (optional)
│           ├── resolverPartials: { ... } (optional)
│           └── tests:
│               ├── Test Case 1.1
│               │   ├── desc: "description_1.1"
│               │   ├── data: { ... } (optional)
│               │   └── expect: { ... }
│               ├── Test Case 1.2
│               │   ├── desc: "description_1.2"
│               │   ├── data: { ... } (optional)
│               │   └── expect: { ... }
│               └── ... (more test cases)
│       └── Test Suite 2
│           ├── name: "suite_name_2"
│           ├── template: "..."
│           └── tests:
│               ├── Test Case 2.1
│               │   ├── desc: "description_2.1"
│               │   └── expect: { ... }
│               └── ... (more test cases)
│       └── ... (more test suites)
│
└── helpers/
    ├── *.yaml (e.g., history.yaml, json.yaml, ifEquals.yaml, etc.)
    │   └── List of Test Suites (same structure as above)
    │       └── Test Suite H1
    │           ├── name: "helper_suite_name_1"
    │           ├── template: "..."
    │           └── tests:
    │               ├── Test Case H1.1
    │               │   ├── desc: "description_H1.1"
    │               │   └── expect: { ... }
    │               └── ...
    └── ... (more helper spec files)
```
"""

from __future__ import annotations

import asyncio
import re
import unittest
from collections.abc import Callable, Coroutine
from pathlib import Path
from typing import Any, Generic, TypedDict

import structlog
import yaml
from pydantic import BaseModel, Field

from dotpromptz.dotprompt import Dotprompt
from dotpromptz.typing import (
    DataArgument,
    JsonSchema,
    Message,
    ModelConfigT,
    PromptInputConfig,
    PromptMetadata,
    ToolDefinition,
)

logger = structlog.get_logger(__name__)


CURRENT_FILE = Path(__file__)
ROOT_DIR = CURRENT_FILE.parent.parent.parent.parent.parent
SPECS_DIR = ROOT_DIR / 'spec'

# List of files that are allowed to be used as spec files.
# Useful for debugging and testing.
ALLOWLISTED_FILES = [
    'spec/helpers/history.yaml',
    'spec/helpers/ifEquals.yaml',
    'spec/helpers/json.yaml',
    'spec/helpers/media.yaml',
    'spec/helpers/role.yaml',
    'spec/helpers/unlessEquals.yaml',
    'spec/variables.yaml',
    # 'spec/helpers/section.yaml',
    # 'spec/metadata.yaml',
    # 'spec/partials.yaml',
    # 'spec/picoschema.yaml',
]

# Counters for test class and test method names.
suite_counter = 0
test_case_counter = 0


class Expect(BaseModel):
    """An expectation for the spec."""

    config: dict[Any, Any] = Field(default_factory=dict)
    ext: dict[str, dict[str, Any]] = Field(default_factory=dict)
    input: PromptInputConfig | None = None
    messages: list[Message] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    raw: dict[str, Any] | None = None


class SpecTest(BaseModel, Generic[ModelConfigT]):
    """A test case for a YAML spec."""

    desc: str = Field(default='UnnamedTest')
    data: DataArgument[Any] | None = None
    expect: Expect
    options: PromptMetadata[ModelConfigT] | None = None


class SpecSuite(BaseModel, Generic[ModelConfigT]):
    """A suite of test cases for a YAML spec."""

    name: str = Field(default='UnnamedSuite')
    template: str
    data: DataArgument[Any] | None = None
    schemas: dict[str, JsonSchema] | None = None
    tools: dict[str, ToolDefinition] | None = None
    partials: dict[str, str] = Field(default_factory=dict)
    resolver_partials: dict[str, str] = Field(default_factory=dict)
    tests: list[SpecTest[ModelConfigT]] = Field(default_factory=list)


def is_allowed_spec_file(file: Path) -> bool:
    """Check if a spec file is allowed.

    Args:
        file: The file to check.

    Returns:
        True if the file is allowed, False otherwise.
    """
    fname = file.absolute().as_posix()
    for allowed_file in ALLOWLISTED_FILES:
        if fname.endswith(allowed_file):
            return True
    return False


def sanitize_name_component(name: str | None) -> str:
    """Sanitizes a name component for use in a Python identifier.

    Args:
        name: The name to sanitize.

    Returns:
        A sanitized name.
    """
    name_str = str(name) if name is not None else 'None'
    name_str = re.sub(r'[^a-zA-Z0-9_]', '_', name_str)
    if name_str and name_str[0].isdigit():
        name_str = '_' + name_str
    return name_str or 'unnamed_component'


def make_test_method_name(yaml_file_name: str, suite_name: str | None, test_desc: str | None) -> str:
    """Creates a sanitized test method name.

    Args:
        yaml_file_name: The name of the YAML file.
        suite_name: The name of the suite.
        test_desc: The description of the test.

    Returns:
        A sanitized test method name.
    """
    file_part = sanitize_name_component(yaml_file_name.replace('.yaml', ''))
    suite_part = sanitize_name_component(suite_name)
    desc_part = sanitize_name_component(test_desc)
    return f'test_{file_part}_{suite_part}_{desc_part}_'


def make_test_class_name(yaml_file_name: str, suite_name: str | None) -> str:
    """Creates a sanitized test class name for a suite.

    Args:
        yaml_file_name: The name of the YAML file.
        suite_name: The name of the suite.

    Returns:
        A sanitized test class name.
    """
    file_part = sanitize_name_component(yaml_file_name.replace('.yaml', ''))
    suite_part = sanitize_name_component(suite_name)
    return f'Test_{file_part}_{suite_part}Suite'


def make_dotprompt_for_suite(suite: SpecSuite[ModelConfigT]) -> Dotprompt:
    """Constructs and sets up a Dotprompt instance for the given suite.

    Args:
        suite: The suite to construct a Dotprompt for.

    Returns:
        A Dotprompt instance.
    """
    resolver_partials_from_suite: dict[str, str] = suite.resolver_partials

    def partial_resolver_fn(name: str) -> str | None:
        return resolver_partials_from_suite.get(name)

    dotprompt = Dotprompt(
        schemas=suite.schemas,
        tools=suite.tools,
        partial_resolver=partial_resolver_fn if suite.resolver_partials else None,
    )

    # Register partials directly defined in the suite
    defined_partials: dict[str, str] = suite.partials
    for name, template_content in defined_partials.items():
        dotprompt.define_partial(name, template_content)

    return dotprompt


class TestSpecFiles(unittest.IsolatedAsyncioTestCase):
    """Runs essential checks to ensure the spec directory is valid."""

    def test_spec_path(self) -> None:
        """Test that the spec directory exists."""
        self.assertTrue(SPECS_DIR.exists())
        self.assertTrue(SPECS_DIR.is_dir())

    def test_spec_path_contains_yaml_files(self) -> None:
        """Test that the spec directory contains YAML files."""
        self.assertTrue(list(SPECS_DIR.glob('**/*.yaml')))

    def test_spec_files_are_valid(self) -> None:
        """Test that all spec files contain valid YAML."""
        for file in SPECS_DIR.glob('**/*.yaml'):
            with open(file) as f:
                data = yaml.safe_load(f)
                self.assertIsNotNone(data)


class YamlSpecTestBase(unittest.IsolatedAsyncioTestCase, Generic[ModelConfigT]):
    """A base class that is used as a template for all YAML spec test suites."""

    async def run_yaml_test(
        self, yaml_file: Path, suite: SpecSuite[ModelConfigT], test_case: SpecTest[ModelConfigT]
    ) -> None:
        """Runs a YAML test.

        Args:
            yaml_file: The path to the YAML file.
            suite: The suite to run the test on.
            test_case: The test case to run.

        Returns:
            None.
        """
        logger.info(f'[TEST] {yaml_file.stem} > {suite.name} > {test_case.desc}')

        # Create test-specific dotprompt instance.
        dotprompt = make_dotprompt_for_suite(suite)
        self.assertIsNotNone(dotprompt)

        data = self._merge_data(suite.data or DataArgument[Any](), test_case.data or DataArgument[Any]())
        result = await dotprompt.render(suite.template, data, test_case.options)
        pruned_res: Expect = Expect(**result.model_dump())
        self.assertEqual(pruned_res, test_case.expect)

    def _merge_data(self, data1: DataArgument[Any], data2: DataArgument[Any]) -> DataArgument[Any]:
        merged = DataArgument[Any]()
        merged.input = data1.input or data2.input
        merged.docs = (data1.docs or []) + (data2.docs or [])
        merged.messages = (data1.messages or []) + (data2.messages or [])
        merged.context = {**(data1.context or {}), **(data1.context or {})}
        return merged


def make_suite_class_name(yaml_file: Path, suite_name: str | None) -> str:
    """Creates a class name for a suite.

    Args:
        yaml_file: The path to the YAML file.
        suite_name: The name of the suite.

    Returns:
        A class name for the suite.
    """
    global suite_counter
    suite_counter += 1
    file_part = sanitize_name_component(yaml_file.stem)
    suite_part = sanitize_name_component(suite_name)
    return f'Test_{file_part}_{suite_part}Suite_{suite_counter}'


def make_test_case_name(yaml_file: Path, suite_name: str, test_desc: str) -> str:
    """Creates a test case name.

    Args:
        yaml_file: The path to the YAML file.
        suite_name: The name of the suite.
        test_desc: The description of the test.

    Returns:
        A test case name.
    """
    global test_case_counter
    test_case_counter += 1
    file_part = sanitize_name_component(yaml_file.stem)
    suite_part = sanitize_name_component(suite_name)
    test_method_part = sanitize_name_component(test_desc)
    return f'test_{file_part}_{suite_part}_{test_method_part}_{test_case_counter}'


def make_async_test_case_method(
    yaml_file: Path,
    suite: SpecSuite[ModelConfigT],
    test_case: SpecTest[ModelConfigT],
) -> Callable[[YamlSpecTestBase[ModelConfigT]], Coroutine[Any, Any, None]]:
    """Creates an async test method for a test case.

    Args:
        yaml_file: The path to the YAML file.
        suite: The suite to create the test method for.
        test_case: The test case to create the test method for.

    Returns:
        An async test method.
    """

    async def test_method(self_dynamic: YamlSpecTestBase[ModelConfigT]) -> None:
        """An async test method."""
        await self_dynamic.run_yaml_test(yaml_file, suite, test_case)

    return test_method


def make_async_skip_test_method(
    yaml_file: Path, suite_name: str
) -> Callable[[YamlSpecTestBase[ModelConfigT]], Coroutine[Any, Any, None]]:
    """Creates a skip test for a suite.

    Args:
        yaml_file: The path to the YAML file.
        suite_name: The name of the suite.

    Returns:
        A skip test.
    """

    async def skip_method(self_dynamic: YamlSpecTestBase[ModelConfigT]) -> None:
        self_dynamic.skipTest(f"Suite '{suite_name}' in {yaml_file.stem} has no tests.")

    return skip_method


def generate_test_suites(files: list[Path]) -> None:
    """Dynamically generates test suite classes and methods from YAML spec files.

    Args:
        files: A list of YAML spec files to generate test suites from.

    Returns:
        None.
    """
    module_globals = globals()

    for yaml_file in files:
        if not is_allowed_spec_file(yaml_file):
            logger.warn('Skipping non-allowlisted spec file for class generation', file=str(yaml_file))
            continue

        # Load the YAML file and ensure it's valid.
        try:
            with open(yaml_file, encoding='utf-8') as f:
                suites_data = yaml.safe_load(f)
            if not suites_data:
                logger.warn('Skipping spec file with no data', file=str(yaml_file))
                continue
        except yaml.YAMLError as e:
            logger.error('Error loading spec file', file=str(yaml_file), error=e)
            raise

        # Iterate over the suites in the YAML file and ensure it has a name.
        for suite_data in suites_data:
            # Normalize the suite data to ensure it has a name.
            suite = SpecSuite(**suite_data)
            suite.name = suite.name or f'UnnamedSuite_{yaml_file.stem}'

            # Create the dynamic test class for the suite.
            class_name = make_suite_class_name(yaml_file, suite.name)
            klass = type(class_name, (YamlSpecTestBase,), {})

            # Skip the suite if it has no tests.
            if not suite.tests:
                klass.test_empty_suite = make_async_skip_test_method(yaml_file, suite.name)  # type: ignore[attr-defined]

            # Iterate over the tests in the suite and add them to the class.
            for tc in suite.tests:
                # Create the test case method and add it to the class.
                test_case_name = make_test_case_name(yaml_file, suite.name, tc.desc)
                test_method = make_async_test_case_method(yaml_file, suite, tc)
                setattr(klass, test_case_name, test_method)

            # Add the test suite class to the module globals.
            module_globals[class_name] = klass


generate_test_suites(list(SPECS_DIR.glob('**/*.yaml')))

if __name__ == '__main__':
    unittest.main()
